storage/innobase/handler/ha_innodb.cc

```diff
@@ -16639,14 +16714,6 @@ ha_innobase::store_lock(
 	}
 	m_prebuilt->skip_locked= (lock_type == TL_WRITE_SKIP_LOCKED ||
 				  lock_type == TL_READ_SKIP_LOCKED);
-
-	if (!trx_is_started(trx)
-	    && (m_prebuilt->select_lock_type != LOCK_NONE
-	        || m_prebuilt->stored_select_lock_type != LOCK_NONE)) {
-
-		trx->will_lock = true;
-	}
-
 	return(to);
 }
```

Comment:
Before I forget: @montywi made a comment about this last week.

I will need to check how `ha_innobase::store_lock()` and `ha_innobase::external_lock()` are being invoked in
```sql
insert into t values(1);
insert into t values(2);
```
Both with `autocommit=0` and `autocommit=1` as well as with explicit `START TRANSACTION` and `COMMIT` and additionally inside `LOCK TABLES`/`UNLOCK TABLES`. Possibly this `trx->will_lock=true` assignment needs to be restored. I had been under the impression that this code is just duplicating something that `ha_innobase::external_lock()` is executing.

Reply:
I tested the following SQL with `./mtr --rr`:
```sql
--source include/have_innodb.inc
create table t(a int primary key) engine=innodb;
insert into t values(1);
insert into t values(2);
set autocommit=0;
insert into t values(3);
insert into t values(4);
commit;
lock tables t write;
insert into t values(5);
insert into t values(6);
commit;
unlock tables;
set autocommit=1;
begin;
insert into t values(7);
insert into t values(8);
commit;
lock tables t write;
begin;
insert into t values(9);
insert into t values(10);
commit;
unlock tables;
start transaction with consistent snapshot;
insert into t values(11);
insert into t values(12);
commit;
start transaction with consistent snapshot;
lock tables t write;
insert into t values(13);
insert into t values(14);
commit;
begin;
lock tables t write;
insert into t values(15);
insert into t values(16);
commit;
drop table t;
```
For each `lock tables` and `unlock tables` as well as for each `insert` statement except (5), (6), (13), (14), (15), (16), the function `ha_innobase::store_lock()` will be invoked once, followed by 2 calls of `ha_innobase::external_lock()`, with `F_WRLCK` and `F_UNLCK`. The first `ha_innobase::external_lock()` would execute the following:
```c++
		if (!trx_is_started(trx)
		    && (m_prebuilt->select_lock_type != LOCK_NONE
			|| m_prebuilt->stored_select_lock_type != LOCK_NONE)) {

			trx->will_lock = true;
		}
```
This basically duplicates the code that I would remove from `ha_innobase::store_lock()`. I added watchpoints on `trx->state` as well as `trx->will_lock` to find out if we might neglect setting the latter. The state changes are as follows:
* (1): `trx->will_lock=true` in `ha_innobase::external_lock()`, transaction started with `ha_innobase::write_row()`, committed in `mysql_execute_command()` (`trx_t::commit_cleanup()` will clear `will_lock=false`)
* (2): ditto
* (3): starts like (1), but there is an explicit `commit` after (4)
* (5) and (6): like (3) and (4), but `trx->will_lock=true` and `trx->state=TRX_STATE_ACTIVE` will be set `lock tables t write` by `row_lock_table()`, invoked by `ha_innobase::external_lock(F_WRLCK)`
* (7) and (8): like (3) and (4)
* (9) and (10): like (3) and (4); `will_lock` will be set and cleared during `lock tables`, which will be auto-committed
* (11) and (12): During the execution `START TRANSACTION WITH CONSISTENT SNAPSHOT`, `trx->will_lock = true` will be set by `trx_start_low()`.
* (13) and (14): Like (1) and (2), that is, 2 separate autocommit `insert`! This was a real surprise. The `LOCK TABLE` implicitly commit the `START TRANSACTION WITH CONSISTENT SNAPSHOT` and then starts and auto-commit itself. For each `insert`, the `trx->will_lock` will be set in `ha_innobase::start_stmt()` inside `open_and_lock_tables()`, and the transaction started in `handler::ha_write_row()` and autocommitted in `mysql_execute_command()`, like (1) and (2).
* (15) and (16): Like (13) and (14), but the initial `BEGIN` will not start any transaction inside InnoDB.

In each and every case we do have `trx->will_lock=true` when the transaction is being started. So, the code in `ha_innobase::store_lock()` is redundant.

The only somewhat surprising finding (unrelated to these changes) is that `LOCK TABLES` can basically only work when it is executed before `BEGIN` or `START TRANSACTION`. It could be good to issue a warning (or an error if some particular `sql_mode` is set) when there `LOCK TABLES` is executed inside a transaction.